### PR TITLE
feat: Debug run navigation

### LINKED ISSUE #25899: Debug Run Navigation
### Current behavior

On the Debug page, the user is only allowed to see the most recent "relevant" run that is determined when querying for runs based on the Git commit shas of that user's local environment. There are use cases where a user would wish to select from other runs.

See "Problem Statement" from ClickUp for further details: https://app.clickup.com/18033298/docs/h6amj-41627/h6amj-36067

### Desired behavior

Allow a user to switch between all "relevant runs".

_New run navigation component_
Screenshot 2023-02-21 at 12 03 40 PM

ClickUp requirements:  https://app.clickup.com/18033298/docs/h6amj-41627/h6amj-36087

#### Technical Requirements
* Refactor/replace `DebugRelevantRunBar` with new navigation component 
  * If there is only one relevant run, the Debug page will be showing that and this component will be hidden
  * Component is made up of
    * Header
      * Shows information about the most recent relevant run.  	
      * Will have a prominent CTA button which says "Switch to latest run"
      * CTA will be enabled if page is not showing the most recent run.  Clicking it will move the page to the most recent run
      * Will show "You are on the most recent run" if the page is showing the most recent run and CTA will be removed
      * Will show a caret on the left to indicate it is expandable if
        * there are any relevant runs older than current 
        * more than 1 relevant run more recent than current
      * UI Concerns
        * Hide commit message and spec counts and smaller breakpoints

**Example Header States**

_Most recent run is running and component is not expandable because there is only one more recent run than selected_
Debug_Run_Navigation_Header_One

_Most recent run is running and component is expandable because there is more than one run that can be switched to_
Debug_Run_Navigation_Header_Expanded_Running

_Most recent run is the one currently shown on the page_
Screenshot 2023-02-22 at 9 01 11 AM

_Most recent run is completed and is not currently selected_
Debug_Run_Navigation_Header_Expanded_Completed

    * Commit list
      * Show list of commits that contain relevant runs returned after querying Cypress Cloud
      * Include the current commit of the user's environment even if it does not have any runs
      * Will be in descending order by run number
      * Will show a message at the bottom if more than 100 runs are found that allows a user to follow a link to the project in Cypress Cloud
      * UI Concerns
        * Should have a max-height of 50%-70% of viewport height and scroll content if it is larger than that

_Complete commit list showing commits with runs_
Screenshot 2023-02-21 at 12 42 54 PM

_Message shown at bottom of the expanded component if more than 100 runs are found_
Screenshot 2023-02-21 at 12 36 45 PM

    * Commit
      * Show commit sha and commit message
      * Will have a "Checked out" badge if the commit is the currently checked out commit
      * Runs associated with the commits will be nested underneath
      * UI Concerns
        * Truncate commit message at smaller breakpoints

_Commit row showing the "Checked out" badge_
Screenshot 2023-02-21 at 12 37 22 PM

    * Runs
      * Show run status component, test counts component, spec counts, and duration/createdAt component
      * Will have a grey background and a checkmark shown if it is the run currently selected by the page
      * Runs that are not the currently selected run should be able to be selected
        * Will show the same grey background on hover for all runs that are not the current selected run (see Figma [prototype](https://www.figma.com/proto/ooXxp2usVdriapxZsFXFhn/In-App-Test-Results-View?page-id=1151%3A122273&node-id=1892%3A178911&viewport=-8026%2C2502%2C0.25&scaling=min-zoom&starting-point-node-id=1892%3A178911&show-proto-sidebar=1))	
        * Should show appropriate cursor to indicate it can be selected
        * Should be keyboard navigable
        * Selecting a row via clicking or with the keyboard should close the expanded component and move the Debug page to show the selected run
        * UI Concerns
          * Hide spec counts and duration/completedAt information at smaller breakpoints

_Shows a selected and non selected run_
Screenshot 2023-02-21 at 12 37 43 PM

### Polling Updates

Currently, there are two pollers that help control the features of the Debug page:

* `RelevantRunsDataSource` controls the call to `runsByCommitSha` to determine the CloudRuns that are determined to be most "relevant" to the list of local commit shas.  It polls every 30 seconds to determine which runs are considered `current` and `next`
* `RelevantRunSpecsDataSource` controls the call to `runsByNumber` for running runs. When a running run is found and being shown on the Debug page, it will poll every 15 seconds to return the updated information for the running run

**Updates**
The new Run Navigation component will now display up to 100 additional runs to pick from instead of just `current` and `next`.  The following updates will be needed to support this functionality

* `RelevantRunsDataSource` should now calculate are return
  * `current` - this remain the same and be the run used by the sidebar badge as well as the Debug page for showing the contents on the main part of the page.  
  * `latest` - this should replace `next` and will be the run used by the Run Navigation in the header part of the component.  If `latest` is the same as `current`, then that is when the 'You are on the latest run' state for the header should be used
  * `all` - this will represent all relevant runs calculated for the given commit shas (including `current` and `latest` above).  The logic should go throw the returned runs and stop when it finds a commit sha that has at least one "completed" run (status of run is not RUNNING).  It should still return all runs for that furthest back commit that meets that criteria [See "What is shown for relevant commits" in the Clickup requirements [here](https://app.clickup.com/18033298/docs/h6amj-41627/h6amj-36087) for more detail]

* `RelevantRunSpecsDataSource` will now possibly need to watch more than 1 running run
  * The initial implementation would always poll for both `current` and `next`, regardless of which run was running. In the issue for showing the details of a running run on the Debug page, this datasource is being updated to have a runNumber as a parameter so that it only polls for the running run.  This will need to be updated to allow for an array of run numbers to poll for that represents all the running runs.  The GQL query in this datasource will need to be generated dynamically to create an alias for each run number that should be polled. Additional fields will need to be added to support the additional data being shown in the new component (test counts, duration, createdAt)

**Relevant run switching and persistence**
The `RelevantRunsDataSource` caches the `current` run that is shown on the Debug page.  The logic for this cache should mostly stay as it is currently set, except for one scenario:
* When determining if the cached `current` run should change, the logic should consider the commit sha.  If a completed run is found that is newer than the `current` run, only update the cached run if the newer run is for a newer(i.e. different) commit.  This will allow a user to remain on an older run within the most recent commit if it was being viewed.

 [See "Relevant run switching and persistence" in the Clickup requirements [here](https://app.clickup.com/18033298/docs/h6amj-41627/h6amj-36087) for more detail]

#### References 
ClickUp - https://app.clickup.com/t/18033298/PM-2859
Figma - https://www.figma.com/file/ooXxp2usVdriapxZsFXFhn/In-App-Test-Results-View?node-id=1151%3A122273&t=ngBnD28gqvAenHrq-1
Figma prototype - [Showing some interactions] https://www.figma.com/proto/ooXxp2usVdriapxZsFXFhn/In-App-Test-Results-View?page-id=1151%3A122273&node-id=1892%3A178911&viewport=-8026%2C2502%2C0.25&scaling=min-zoom&starting-point-node-id=1892%3A178911&show-proto-sidebar=1

### LINKED ISSUE #26018: Debug Run Navigation: Frontend Updates
Build the UI component and associated states in https://github.com/cypress-io/cypress/issues/25899

You will leave heavily on GraphQL for data; the query will be based on https://github.com/cypress-io/cypress/blob/79b2c38e3546516b4c74f40016fff3669b6e9073/packages/app/src/debug/DebugContainer.vue#L154-L157 with one addition, an `allRuns` property that contains and array of `CloudRun`.

```graphql
gql`
fragment DebugSpecs on Query {
  currentProject {
    id
    cloudProject {
      __typename
      ... on CloudProject {
        id
   
        # This is the additional field
        allRuns: runsByCommitShas(commitShas: $commitShas) {
          id
          # whatever you need
        }
      }
    }
  }
}
```

Note that `allRuns` is an array containing various runs and associated with different git hashes; you'll need to do some post-processing to group the runs by git hash.

There's a few different states to consider; these are detailed in the original ticket, https://github.com/cypress-io/cypress/issues/25899, and in Figma (see "Resources" in that ticket).

### PR DESCRIPTION
**Pull request by lmiller1990**
- Closes https://github.com/cypress-io/cypress/issues/25899
- Closes https://github.com/cypress-io/cypress/issues/26018

### Additional details

There are two main parts to this PR:
1. addition of the new `DebugRunNavigation` component and its child `DebugRunNavigationRow` component
2. Updates to the `RelevantRunsDataSource` and `RelevantRunSpecsDataSource` to support it

The other changes all involve refactoring various parts of the code for reuse, and some general UI fixes/improvements that were found will testing this feature.

### Steps to test

You can see various states of this component by opening the App for existing projects (like our own Cypress project), but the easiest way to test will be to have your own Cloud project that you can easily record local runs against.

Scenarios to test:
- Start with a project with no runs.  When connected to the cloud you should see the existing empty state to record a run
- Record the first run in the project, the Debug page should update to show the running build and no run navigation should be present since it is the only run
- Make a commit in your project and then record a new run
- The debug page should update to show the run navigation component that indicates that the new build is running
image
  - The component will **not** show the expandable component since there is only one run to switch to
  - Click the "Switch to the newest run" button to view the running build
  - The component will update to show you are on the most recent build and you should be able to expand the component and switch back to the previous running build
image

- When the running build completes, the component will be removed since there is only one build for the most recent commit with a completed build and no newer builds
- Record another run **without** making a new commit
  - The run navigation component will appear again and show the running build
  - You can switch back and forth between the runs 
- When the running build completes, the navigation component will remain.  This is because there are multiple builds for the same most relevant commit with builds.  

### How has the user experience changed?

The Debug page will now show a new component at the top when scenarios allow for changing between different runs for debugging.

See the linked issue https://github.com/cypress-io/cypress/issues/25899 for screenshots and details.

### PR Tasks

- [x] Have tests been added/updated?
- [] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

File: packages/app/src/debug/DebugRunDetailedView.vue
Comment (warrensplayer) [inline packages/app/src/debug/DebugRunDetailedView.vue]: What is the need for the "ring" styles here?  It causes the dashed line to get broken up.

File: packages/app/src/debug/DebugRunDetailedView.vue
Comment (warrensplayer) [inline packages/app/src/debug/DebugRunDetailedView.vue]: Instead of adding a DIV here for the dashed line which is not semantically correct inside a `ul`, you can add these same classes to a `before` pseudo element on the `ul` above.  

It would be something like:
```
before:(content-DEFAULT h-full w-5px border-2 border-dashed border-l-0 border-y-0 border-r-gray-100 left-[15px] absolute)
```

Also, one option to have it match the design more closely would be to remove the `h-full` and use a `top` and `bottom` offset instead.
```
before:(content-DEFAULT top-20px bottom-10px w-5px border-2 border-dashed border-l-0 border-y-0 border-r-gray-100 left-[15px] absolute)
```

File: packages/app/src/debug/DebugRunDetailedView.vue
Comment (warrensplayer) [inline packages/app/src/debug/DebugRunDetailedView.vue]: This should be comparing run numbers, not shas here. 
```suggestion
  return latest.value?.runNumber === current.value?.runNumber
```

File: packages/app/src/debug/DebugRunDetailedView.vue
Comment (warrensplayer) [inline packages/app/src/debug/DebugRunDetailedView.vue]:
```suggestion
        <Button
          v-if="!latestIsCurrentlySelected"
          data-cy="switch-to-latest"
          @click="$event => changeRun(latest!)"
        >
```
Just adding this one click event should get this button working

**Review by warrensplayer — COMMENTED** [review]
@lmiller1990 I wanted to get you this feedback.  When a new build is running, I am getting GQL errors and still looking into what is causing that.

File: packages/app/src/debug/DebugRunDetailedView.vue
Comment (lmiller1990) [reply packages/app/src/debug/DebugRunDetailedView.vue]: I added this to get this gap: 
Screenshot 2023-03-13 at 9 23 09 am

Definitely room to improve the CSS.

File: packages/app/src/debug/DebugRunDetailedView.vue
Comment (lmiller1990) [reply packages/app/src/debug/DebugRunDetailedView.vue]: Thanks!

File: packages/data-context/src/sources/RelevantRunsDataSource.ts
Comment (lmiller1990) [inline packages/data-context/src/sources/RelevantRunsDataSource.ts]: We should just make `allRuns` into a `Set`, then do `[...allRuns]` at the end - this would ensure no duplicates, and avoid this conditional entirely 😇 

I will add this in soon.

File: packages/data-context/src/sources/RelevantRunsDataSource.ts
Comment (lmiller1990) [inline packages/data-context/src/sources/RelevantRunsDataSource.ts:1]: ## Note for reviewers

Above file `packages/data-context/src/sources/RelevantRunSpecsDataSource.ts` is "Large diffs are not rendered by default" but it's a really critical part of this PR - make sure you expand it and take a look!

File: packages/data-context/src/sources/RelevantRunSpecsDataSource.ts
Comment (lmiller1990) [inline packages/data-context/src/sources/RelevantRunSpecsDataSource.ts:130]: When is `run.id` not going to be equal to `runId`? It looks like this is what  causes 

```js
    function subscribed (value: any) {
      //optional filter for stream of data
      if (opts?.filter && !opts.filter(value)) {
        return
      }

      // We can get events here before next() is called setting up the deferred promise
```

To return early - but I think I'm still missing what this *does* - I'm fairly sure it's to unsubscribe the subscription when the run is no longer running, but I am still figuring out the entirely flow.

File: packages/data-context/src/sources/RelevantRunSpecsDataSource.ts
Comment (lmiller1990) [inline packages/data-context/src/sources/RelevantRunSpecsDataSource.ts:139]: This is pretty clean - it looks like unless someone has to change something significant, this code should be set and forget - it won't really need to be revisited, which is great! 

I'm still not 100% sold on dynamically generating the query, since we lose the declarative nature of GraphQL, which is really nice (you can just copy+paste the code into the Graphical interface and run the query, no longer possible here) but I do think this is the best option we have right now.

File: packages/data-context/src/sources/RelevantRunSpecsDataSource.ts
Comment (lmiller1990) [inline packages/data-context/src/sources/RelevantRunSpecsDataSource.ts:232]: Nice job figuring this out 💯

File: packages/data-context/test/unit/sources/fixtures/graphqlFixtures.ts
Comment (lmiller1990) [inline packages/data-context/test/unit/sources/fixtures/graphqlFixtures.ts:144]: Not reltaed to this file but I found out that

image

When I am on the **spec page** this will show the old status until the new run is completed done - eg, if my run takes 10 min, even if the first test fails, I won't know until the entire run finishes. Is that correct, or should this start auto upadating?  Potential bug in polling logic for the Sidebar subscription?

The updates on the debug page are working as expected.

**Review by lmiller1990 — COMMENTED** [review]
Still testing and going over the code, seems mostly working as expected so far, left some comments/questions.

File: packages/data-context/test/unit/sources/fixtures/graphqlFixtures.ts
Comment (warrensplayer) [reply packages/data-context/test/unit/sources/fixtures/graphqlFixtures.ts:144]: > Not reltaed to this file but I found out that
> 
> image
> 
> When I am on the **spec page** this will show the old status until the new run is completed done - eg, if my run takes 10 min, even if the first test fails, I won't know until the entire run finishes. Is that correct, or should this start auto upadating? Potential bug in polling logic for the Sidebar subscription?
> 
> The updates on the debug page are working as expected.

This is the expected behavior at this time.  There is another initiative called [In-App Run Notifications](https://app.clickup.com/t/18033298/PM-3555) to address providing better feedback to the user about Cloud runs when you are not on the Debug page or not looking at the Cypress app at all.

File: packages/data-context/src/sources/RelevantRunSpecsDataSource.ts
Comment (warrensplayer) [reply packages/data-context/src/sources/RelevantRunSpecsDataSource.ts:130]: I have added some JSDocs to the `DataEmitterActions.subscribeTo` method to better describe the options. This filter is a predicate to allow for filtering the values being "emitted" for a subscription.  Since there can be multiple "subscriptions" tied to the same "event", I needed a way to only send certain values back on certain subscriptions.  

In the Run Navigation use case, there could be multiple running builds in the Cloud at one time that the component wants to subscribe to.  The backend query will combine them together and execute one query to the Cloud using the `cloudNodesById` field.  When the result comes back, the code iterates over each `CloudRun` returned to check the cache to see if it changes.  If it has changed and an event is sent to the emitter, the `filter` will only send through the runs with the matching `id` to the correct upstream (i.e. front end) subscription.

**Comment by marktnoonan** [comment]
mostly works great, found one small bug with the time incrementing incorrectly after runs complete in some situations: https://www.loom.com/share/5cd609f2ff59433e8c9b8795cc1ba957?highlightComment=11609224&t=435

That video is really long so I'll gather up all the feedback that was actually useful and add it here, mostly around a11y, and a little bit around the legibility of the UI, nothing blocking apart from the bug.

File: packages/app/src/debug/DebugRunNavigationRow.vue
Comment (marktnoonan) [inline packages/app/src/debug/DebugRunNavigationRow.vue:8]: This aria label has some problems. I think there needs to be a `computed` for the `runNumber` to populate, as it's probably just evaluated once on render, meaning we have an empty run number in the label.

Screenshot 2023-03-27 at 1 55 18 PM

But when this renders a button, the presence of `aria-label` prevents the internal content from being exposed to screen readers, the label becomes the accessible name. When it's a div and not interactive this doesn't seem to be a problem. [Loom](https://www.loom.com/share/5cd609f2ff59433e8c9b8795cc1ba957?highlightComment=11609758&t=939).

Comment (marktnoonan) [reply packages/app/src/debug/DebugRunNavigationRow.vue:8]: A click hander on a div is fine if there is a button inside that can emit a click and be the focus target for keyboard users, even if the visual style makes the whole row look selected.

File: cli/CHANGELOG.md
Comment (astone123) [inline cli/CHANGELOG.md]:
```suggestion
- The [Debug page](https://docs.cypress.io/guides/cloud/runs#Debug) now allows for navigating between all runs recorded for a commit. Addresses [#25899](https://github.com/cypress-io/cypress/issues/25899) and [#26018](https://github.com/cypress-io/cypress/issues/26018).
```

Comment (mike-plummer) [inline cli/CHANGELOG.md]: I think Jess is planning a blog post for this - I think we typically include links to those from the changelog IIRC. Add link now or just make sure person doing the release knows to add when they prep the changelog

File: packages/app/src/debug/DebugCommitIcon.vue
Comment (mike-plummer) [inline packages/app/src/debug/DebugCommitIcon.vue:2]: Not worth holding up this PR, but should this be added to the design system as a shared icon?

**Comment by warrensplayer** [comment]
> found one small bug with the time incrementing incorrectly after runs complete

Looking into this, but unable to replicate in the tests yet.

File: packages/app/src/debug/useDebugRunSummary.ts
Comment (mike-plummer) [inline packages/app/src/debug/useDebugRunSummary.ts]: Remove?

File: packages/app/src/debug/useDebugRunSummary.ts
Comment (mike-plummer) [inline packages/app/src/debug/useDebugRunSummary.ts]: Remove?

**Review by mike-plummer — APPROVED** [review]
Looks good in my local testing, going to approve assuming items already identified will be resolved/answered. Just seeing some cleanup items

File: packages/app/src/debug/useDebugRunSummary.ts
Comment (warrensplayer) [reply packages/app/src/debug/useDebugRunSummary.ts]: Gone

File: packages/app/src/debug/useDebugRunSummary.ts
Comment (warrensplayer) [reply packages/app/src/debug/useDebugRunSummary.ts]: Gone

File: packages/app/src/debug/DebugCommitIcon.vue
Comment (warrensplayer) [reply packages/app/src/debug/DebugCommitIcon.vue:2]: I will file a follow up issue

File: packages/app/src/debug/DebugRunNavigationRow.vue
Comment (warrensplayer) [reply packages/app/src/debug/DebugRunNavigationRow.vue:8]: I follow all of your feedback. I will create a follow up issue for this.  I can see where we also make use of the HeadlessUI radio group component to get better keyboard navigation in that page you mention in the loom.

**Comment by warrensplayer** [comment]
@marktnoonan I fixed the issue with the time counting up not stopping when the run completes that you noted in your loom.  Can you give it another test?

**Review by astone123 — APPROVED** [review]
Everything looked good in my local testing. nice 🎉

File: cli/CHANGELOG.md
Comment (warrensplayer) [reply cli/CHANGELOG.md]: I reached out about this.

File: packages/app/src/debug/DebugContainer.vue
Comment (lmiller1990) [inline packages/app/src/debug/DebugContainer.vue:197]: Hadn't seen this to extract the prop types, neat trick!

File: packages/app/src/pages/Debug.vue
Comment (lmiller1990) [inline packages/app/src/pages/Debug.vue:79]: Should we remove this? Surprised linting is okay.

**Review by lmiller1990 — COMMENTED** [review]
✅

**Pull request merged**
